### PR TITLE
Add RetryableTransaction, update RetryableQuery to prevent database deadlocks in EntityWriteGateway

### PR DESCRIPTION
**Pull request by hanneswernery**
### 1. Why is this change necessary?

As discussed with @OliverSkroblin when writing entities in multiple threads and/or transactions at the same time deadlocks may occur in the database. When a deadlock occurs within a transaction, MySQL rolls back the transaction internally. Therefore the whole transaction can be retried. This is also stated in the MySQL error message:
```
ERROR 1213 (40001): Deadlock found when trying to get lock; try restarting transaction
```
The `EntityWriteGateway` uses a transaction but uses a `RetryableQuery` _inside_ this transaction ([code](https://github.com/shopware/platform/blob/f094b8611d75ac3ca57167b6215cb81da2cafa45/src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php#L104-L106)). In case of a deadlock a single query of the transaction is now retried ([code](https://github.com/shopware/platform/blob/trunk/src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php#L43-L52)) even though the transaction was already rolled back internally by the database. Since queries inside a multi-query-transaction may depend on each other, this causes all kinds of unintended behaviour (i.e. foreign key constraint violations).
The description of the `RetryableException` that is caught also [suggest to retry the transaction,](https://github.com/doctrine/dbal/blob/715a39150cbf961d7ce19b3f1783f6a76a01afe5/lib/Doctrine/DBAL/Exception/RetryableException.php#L8) not a single query.
Therefore the usage of the current `RetryableQuery` only works when running a single query in no transaction.

### 2. What does this change do, exactly?

This PR adds a `RetryableTransaction` class which can be used to execute a number of commands (queries) inside a transaction which can be rolled back and retried as a whole without side effects or unwanted behaviour.
Additionally the current usages of `RetryableQuery` are updated to check whether or not the query is executed inside a transaction. Now if this is the case, the `RetryableQuery` will **not** be retried when a deadlock occurs.
For that the functions `execute` and `retryable` have been backwards-compatibly updated to accept a `Doctrine\DBAL\Connection $connection` to ensure this new safe behaviour.

### 3. Describe each step to reproduce the issue or behaviour.

Run two complex processes, e.g. create new versions of an existing order and update many products at the same time.
A real life example is the order document creation via the administration. This creates a new order version and triggers a complex product subscriber from the Pickware ERP plugin. Since this is a race condition error, your results may vary.

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md
Comment (OliverSkroblin) [inline changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md]: please use the correct keywords:

* Added
* Removed
* Changed
* Deprecated

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (OliverSkroblin) [inline src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]: why not injecting the connection as `__construct` parameter, for the case that we have an instance

**Review by OliverSkroblin — CHANGES_REQUESTED** [review]

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (hanneswernery) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]: That would work as well.

We thought it would be easier when the "deprecation and different parameters" situation is contained in a single method. Otherwise the optional parameter in the constructor would affect a second method and both have to be updated when removing the deprecation 🤷

If you prefer it this way, just says so - I can easily refactor this.

**Comment by hanneswernery** [comment]
I updated the changelog.

More importantly: I updated the `EntityWriteGateway` so it actually uses the new `RetryableTransaction` - which is described in the PR description and was the main reason for this PR. I lost these code changes somehow before opening the PR.

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (mitelg) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]: the constructor is not affected by backward compatibility. if this is your fear 😉

Comment (hanneswernery) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]: > not affected by backward compatibility

How so? 🤔 It's a public class so I can't just change its contract, can I?
I can make the `Connection` constructor parameter optional which moves the parameter checks and deprecation warning out of the `execute` method into the constructor. If you prefer this version.

Either way the goal should be to deprecate the old "`Connection`-less" usages, so with the next major version `RetryableQuery` will always run with a `Connection` instance to make it deadlock safer.

Comment (mitelg) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]: *in _our_ BC strategy 😁

Comment (OliverSkroblin) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]: I'm sorry if I have to prove you wrong, @mitelg, but that only applies to services in the DI container. 
For classes created at runtime, this is a break.

```
$update = new RetryableQuery(
   $this->connection->prepare('UPDATE promotion SET order_count = :count, orders_per_customer_count = :customerCount WHERE id = :id')
);
```

@hanneswernery I see what you mean. But from the semantic point of view, it would be nicer to pass it in the `__construct`. Can you please change this?

File: changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md
Comment (OliverSkroblin) [inline changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md:8]: Sorry I saw it only now. 

Since your change means that plugins have to adapt their source code, please maintain the `UPGRADE INFORMATION` part of the changelog file.

Here is just an example what people have to look for and how it has to look before and after.

**Comment by hanneswernery** [comment]
I added the UPGRADE INFORMATION part of the changelog.
I changed the `RetryableQuery` constructor as discussed.
I rebased the branch onto `trunk`.

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (mitelg) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]: sorry, didn't know this is not a service from the container :smile:

**Comment by J-Rahe** [comment]
Hey @hanneswernery ,
sorry that it took so long to get back to this.
I just checked out your PR and wanted to do the usual rebase, but there are a few merge conflicts.
Could you rebase your PR and fix these conflicts? Then I can put it into our internal review and hopefully get it done soon : )

**Comment by J-Rahe** [comment]
thanks : )

File: changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md
Comment (J-Rahe) [inline changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md:3]: @hanneswernery, you are [missing the author and author_email field](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md). 
If you want you can just send them to me and i add them together with the issue-key. (you can just use your github email, sorry that i didn't notice that earlier, i thought these fields were there : /)

Comment (hanneswernery) [reply changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md:3]: You are right. I just added the necessary meta data to the changelog.
You only have to add the respective issue key 🙂

File: changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md
Comment (windaishi) [inline changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md]:
```suggestion
* Deprecated constructor argument usage `(Doctrine\DBAL\Driver\Statement $query)` in `src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php`.
```

File: changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md
Comment (windaishi) [inline changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md]:
```suggestion
  Use arguments `(Doctrine\DBAL\Connection $connection, Doctrine\DBAL\Driver\Statement $query)` instead.
```

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (windaishi) [inline src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]:
```suggestion
                'Use constructor arguments (Doctrine\DBAL\Connection $connection, Doctrine\DBAL\Driver\Statement $query) instead.',
```

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (windaishi) [inline src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php]:
```suggestion
                'Constructor arguments must be of type (Doctrine\DBAL\Connection $connection, Doctrine\DBAL\Driver\Statement $query).'
```

**Review by windaishi — CHANGES_REQUESTED** [review]

File: changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md
Comment (J-Rahe) [inline changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md]: Please add some before/after example code.

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (J-Rahe) [inline src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php:25]: please add typehints

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (J-Rahe) [inline src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php:54]: please add type hints

**Review by J-Rahe — CHANGES_REQUESTED** [review]
Hey @hanneswernery, i got some feedback during our internal review, and there are a few tests failing. Could you please take a look at that?

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (hanneswernery) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php:25]: If you look at Line 23-35 of this method (https://github.com/shopware/platform/pull/1668/files#diff-bc250d3f5ac62b2d692b9500e6a542b71071ec887778f8d45de93e89a2431d06R23-R35) it is implemented to accept different parameter types.

I chose this implementation to allow a non-breaking, low-refactoring transition from the method interface
```php
public function __construct(Statement $query)
```
to
```php
public function __construct(Connection $connection, Statement $query)
```

If I do not do it this way, I would have needed to create a new class and new methods and deprecate the old ones. I also would have to think of new class and method names which would barely be an improvement to the existing ones (which fit nicely).

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (hanneswernery) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php:54]: See my other comment about this method contract

File: changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md
Comment (hanneswernery) [reply changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md]: I added example usages and old/new usages 👍

File: src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php
Comment (mitelg) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php:25]: if we can not use native typehints, we should at least annotate the types correctly. this would be better as no typehints at all.

Comment (hanneswernery) [reply src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php:25]: I just added the respective (new and correct) PHPDoc type hints.

File: changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md
Comment (J-Rahe) [reply changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md]: @hanneswernery sorry, i should have said that a bit more specific/set the comment a bit lower.
The examples should go into the upgrade information.

Comment (hanneswernery) [reply changelog/_unreleased/2021-02-18-add-retryable-transaction-to-prevent-deadlocks.md]: I moved the the example usages down into the `Upgrade information` section 👍

**Comment by UlrichThomasGabor** [comment]
What is the status of this PR? The faulty handling of deadlocks seems to cause other problems, e.g. https://issues.shopware.com/issues/NEXT-15805

**Comment by UlrichThomasGabor** [comment]
> What is the status of this PR? The faulty handling of deadlocks seems to cause other problems, e.g. https://issues.shopware.com/issues/NEXT-15805

Since nothing happened here for a while I prepared a hotfix from this PR and applied it to our installation. The mentioned issue is not fixed, so it seems something else regarding transaction handling is broken besides this PR.

**Comment by UlrichThomasGabor** [comment]
Here are two `beginTransaction` calls where the transaction code is not wrapped in a try-catch block. Don't know if this is the cause, but it looks wrong to me either way:
- https://github.com/shopware/platform/blob/ab7223a804311d1c2a49dde2bb5dcd5dba1ecb61/src/Core/Content/Product/DataAbstractionLayer/ProductIndexer.php#L158
- https://github.com/shopware/platform/blob/ab7223a804311d1c2a49dde2bb5dcd5dba1ecb61/src/Core/Content/Category/DataAbstractionLayer/CategoryIndexer.php#L166

I guess these could be wrapped in the new `RetryableTransaction` class as well.

**Comment by UlrichThomasGabor** [comment]
These ain't working: (8289d011a9b0f7b1892b5671f0f7eb825ff5ce19, 9da6473672e07f54a49f4252990f3a893327e1de @OliverSkroblin)
- https://github.com/shopware/platform/blob/ab7223a804311d1c2a49dde2bb5dcd5dba1ecb61/src/Core/Framework/DataAbstractionLayer/Indexing/ChildCountUpdater.php#L44
- https://github.com/shopware/platform/blob/ab7223a804311d1c2a49dde2bb5dcd5dba1ecb61/src/Core/Content/Product/DataAbstractionLayer/SearchKeywordUpdater.php#L191
- https://github.com/shopware/platform/blob/ab7223a804311d1c2a49dde2bb5dcd5dba1ecb61/src/Core/Content/Product/DataAbstractionLayer/ProductCategoryDenormalizer.php#L93

These lines catch exceptions which are needed for the control flow of transactions, i.e. which have to reach the new `RetryableTransaction` class.

It also does not make sense to split up large queries for deadlocks. This code is (or should be) wrapped in one transaction either way and InnoDB decides based on the size of the transaction, i.e. how many rows are modified, which one to roll back. Not based on the number of queries.

Best solution here is using `RetryableQuery`, which works with the new `RetryableTransaction` class and which works when the Updater is not wrapped in this class.


Fixing all mentioned issues (wrap ProductIndexer and CategoryIndexer in RetryableTransaction and removing the catch blocks, i.e. remove the split updates) resolves [NEXT-15805](https://issues.shopware.com/issues/NEXT-15805).

**Comment by UlrichThomasGabor** [comment]
By the way, I don't see why the sleep call should be necessary:
https://github.com/shopware/platform/pull/1668/files#diff-104cca720ab1150a9e4052ed394ce25bce03969849e2aeec954156e41189ba21R41
If a deadlock occurs, only one transaction is rolled back. The other one is committed. Likely, when the failing one is retried, the other one has already run through.

I guess in most cases the sleep statement will only prolong the execution of the script. If a transaction is long-running, it might help. But transactions should not be long running anyway and I doubt that a few microseconds will help in that case.

IMHO it would be better to leave the sleep out at the moment and observe how it works out in practice.

**Comment by windaishi** [comment]
@UlrichThomasGabor Thanks for your feedback, we will have a look at this internally. Unfortunately we also are waiting for Shopware to merge this. I will ask them again, if they can merge this PR.

**Comment by UlrichThomasGabor** [comment]
@hanneswernery Why do you think it makes a difference if queries are executed as batch or non-batch inside of a transaction?

From my understanding it should not make a difference:
> It also does not make sense to split up large queries for deadlocks. This code is (or should be) wrapped in one transaction either way and InnoDB decides based on the size of the transaction, i.e. how many rows are modified, which one to roll back. Not based on the number of queries.

Or more specific: From my understanding, performing such work inside of a transaction non-batchy produces only SQL overhead, but will not result in less deadlocks.

**Comment by hanneswernery** [comment]
Hi @UlrichThomasGabor , my last changes are the result of a discussion with @OliverSkroblin. The original PR caused some unwanted behaviour changes on the current `trunk`.

Main reason is that regular non-retryabe exceptions were not mapped (translated into a user-facing error) anymore. This is because [this mapping is only called on non-batchwise executions](https://github.com/shopware/platform/blob/8e56c285549ee9019a6cebffb15af4262361ef79/src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php#L230-L247). The original Shopware code retried _every_ exception non-batchwise manually.
The original PR and the somewhat more correct retry logic that does not retry non-retryable errors (non-batchwise) broke this mapping behaviour.
The last commit reintroduces this manual retry which in turn calls the mapping again. The current state is still pending, I am waiting for a response from @OliverSkroblin

**Comment by UlrichThomasGabor** [comment]
@OliverSkroblin Why is this behavior unwanted?

Basically, this method makes no sense to me:
https://github.com/shopware/platform/blob/8e56c285549ee9019a6cebffb15af4262361ef79/src/Core/Framework/DataAbstractionLayer/Dbal/EntityWriteGateway.php#L116-L146
This is all wrapped in one transaction, so should be executed atomically. Wether, this is executed as a single or as multiple SQL statements, has no effect on whether a deadlock will occur or not. It just prolongs the execution by generating overhead and thus conflicts with the most-basic best practice that transactions should be as fast/small as possible.

If the intend of "batch vs. non-batch" is "should be executed as one transaction or as multiple", then the solution should be to split the transaction or not. That does not happen at the moment.

If the intend of "batch vs. non-batch" is "should block other transaction as least as possible", then changing the isolation level is the way to go. (If this is possible. Otherwise, it becomes complicated...)

Either way, the last commit does not make it worse, but in my opinion does not solve any problem. I would revert it, remove any non-batch code, and if this performs bad in practice, then check if isolation levels can be lowered for certain transactions.

I don't get why batch vs. non-batch has anything to do with throwing Exceptions. This sounds like an unrelated issue to me.

**Comment by OliverSkroblin** [comment]
Hi Ulrich,

first of all, thank you for looking so deeply into this topic.
As you probably noticed I'm not very active on github, so feel free to ping me in the community chat ;). 

Regarding your questions and suggestions:

1) https://github.com/shopware/platform/pull/1668#issuecomment-895197619

This has brought some performance optimizations, and it also happened more often that these places ran into a deadlock or unique constraints. So I added the try catch here to try an single insert/update/delete first and otherwise do a single update/insert.
Of course it can be that this now collides with the new RetryableTransaction.

2) https://github.com/shopware/platform/pull/1668#issuecomment-895961473

The sleep call is necessary, because there could be not only 2 but also 10 transactions that are trying out the same records. These run then at the same time into a deadlock and will try at the same time also again an update. This sleep call is just to make sure that this happens a little bit staggered.

3) https://github.com/shopware/platform/pull/1668#issuecomment-901846298

Perhaps there is a misunderstanding here. The behavior is not unintentional, but rather breaking. We have functions in the core that rely on the event thrown there in the try catch. By changing the entity gateway for the first time the pipeline did not run anymore and tests for duplicate constraints and error reportings were red. Therefore, I played this information back.

However, if this error reporting causes us to have a performance as well as stability loss on the transactions, then I am happy to replace this with an alternative in the next major. 

Only as it is now I can't take it into the core because:
1) It's a breaking change because existing mechanics will be undermined.
2) The pipelines of us are red.

**Comment by hanneswernery** [comment]
As discussed with @OliverSkroblin , I will take another look at the PR and will _remove_ the "unnecessary non-batch manual retry" (😅) again. Instead I will refactor the error mapping a bit so it will work with batchwise executions.

With these changes, this PR will add the correct retry logic _without_ breaking the error mapping and without unnecessary retries 👍

**Comment by UlrichThomasGabor** [comment]
> Hi Ulrich,
> 
> first of all, thank you for looking so deeply into this topic.
> As you probably noticed I'm not very active on github, so feel free to ping me in the community chat ;).

What is the community chat? The forum?


> Regarding your questions and suggestions:
> 
> 1. [Add RetryableTransaction, update RetryableQuery to prevent database deadlocks in EntityWriteGateway #1668 (comment)](https://github.com/shopware/platform/pull/1668#issuecomment-895197619)
> 
> This has brought some performance optimizations, and it also happened more often that these places ran into a deadlock or unique constraints. So I added the try catch here to try an single insert/update/delete first and otherwise do a single update/insert.
> Of course it can be that this now collides with the new RetryableTransaction.

Yes, this conflicts in general with transactions.

This is "more performant" because transactions are shorter and therefore are less likely to deadlock. But this "improvement" comes at a cost, i.e. one indexing action is split into multiple transactions and when an exception occurs, the previous transactions are already committed and the current one fails, thus bringing inconsistency into the data.

Additionally, when the control flow is coming from EntityGateway and it already is in a transaction, this "improvement" is not possible at all.

There are various ways to improve the situation. We can just call, if you want to know my opinion. Might be better than discussing in text form.


> 1. [Add RetryableTransaction, update RetryableQuery to prevent database deadlocks in EntityWriteGateway #1668 (comment)](https://github.com/shopware/platform/pull/1668#issuecomment-895961473)
> 
> The sleep call is necessary, because there could be not only 2 but also 10 transactions that are trying out the same records. These run then at the same time into a deadlock and will try at the same time also again an update. This sleep call is just to make sure that this happens a little bit staggered.

Well, it can stay, it's not seriously problematic, but personally I think it also works when it's not there.


> 1. [Add RetryableTransaction, update RetryableQuery to prevent database deadlocks in EntityWriteGateway #1668 (comment)](https://github.com/shopware/platform/pull/1668#issuecomment-901846298)
> 
> Perhaps there is a misunderstanding here. The behavior is not unintentional, but rather breaking. We have functions in the core that rely on the event thrown there in the try catch. By changing the entity gateway for the first time the pipeline did not run anymore and tests for duplicate constraints and error reportings were red. Therefore, I played this information back.
> 
> However, if this error reporting causes us to have a performance as well as stability loss on the transactions, then I am happy to replace this with an alternative in the next major.
> 
> Only as it is now I can't take it into the core because:
> 
> 1. It's a breaking change because existing mechanics will be undermined.
> 2. The pipelines of us are red.

Yes, this might require more than one step. But the current code, and also the code after merging this PR, is just not working correctly. It's either failing more often than necessary, or data consistency can be lost with a deadlock in the worst case.

**Comment by mitelg** [comment]
> What is the community chat? The forum?

https://slack.shopware.com/ :wink:

**Comment by UlrichThomasGabor** [comment]
@hanneswernery Looks good (to me). Should be squashed I guess.

**Pull request closed without merging**
